Improve compatibility of cast(timestamp as string) with spark

lwz9103 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

In Gluten ClickHouse, Spark's timestamp type is mapped to ClickHouse's datetime64(6) type.
 When casting timestamp '2012-01-01 00:11:22' as a string, Spark returns '2012-01-01 00:11:22', while Gluten ClickHouse returns '2012-01-01 00:11:22.000000'.  

This may lead to incorrect results in some filtering scenarios. Therefore, I added an improvement to remove trailing zeros with the configuration `date_time_64_output_format_cut_trailing_zeros_align_to_groups_of_thousands`, which defaults to `false` to maintain consistency with previous behavior.

If config is true, output scale will be dynamic set in (0, 3, 6) according to tailing zeros. See changes in following examples.
  - 2012-01-01 00:11:22.000000 -> 2012-01-01 00:11:22
  - 2012-01-01 00:11:22.120000 -> 2012-01-01 00:11:22.120
  - 2012-01-01 00:11:22.123400 -> 2012-01-01 00:11:22.123400

Scale values of (0, 3, 6) are the most common in practical scenarios, ensuring that it works in the majority of cases.

Linked issue https://github.com/apache/incubator-gluten/issues/7032

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=69179) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=69179) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=69179) it.</sub>

taiyang-li commented:
@alexey-milovidov can you review it ? Thanks!

baibaichen commented:
@alexey-milovidov, would you please review it? Thanks?

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #69179 from lwz9103/ClickHouse:master -->
*This is an automated comment for commit 7f10b21528cde0d1db7fbdefcb81d6843b81b146 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69179/7f10b21528cde0d1db7fbdefcb81d6843b81b146/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/1c839ff4b95570cb324cfd9307a0825c32fc50c7/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

Algunenano commented on src/Core/Settings.h:
```diff
@@ -1161,6 +1161,7 @@ class IColumn;
     M(DateTimeInputFormat, date_time_input_format, FormatSettings::DateTimeInputFormat::Basic, "Method to read DateTime from text input formats. Possible values: 'basic', 'best_effort' and 'best_effort_us'.", 0) \
     M(DateTimeOutputFormat, date_time_output_format, FormatSettings::DateTimeOutputFormat::Simple, "Method to write DateTime to text output. Possible values: 'simple', 'iso', 'unix_timestamp'.", 0) \
     M(IntervalOutputFormat, interval_output_format, FormatSettings::IntervalOutputFormat::Numeric, "Textual representation of Interval. Possible values: 'kusto', 'numeric'.", 0) \
+    M(Bool, datetime64_trim_suffix_zeros, false, "Trim suffix zeros when datetime64 cast to string", 0) \
```
Making it dynamic to 0, 3, 6 seems completely arbitrary and inconsistent. What's the reasoning for it?

lwz9103 replied:
> The fact that the algo trims to the multiple of 3 show be reflected here or somewhere in documentation as you reflected it in this PR description. Also you need to put a piece of information about this setting to the SettingsHistory

Thanks very much， I will update later.

lwz9103 replied:
> Making it dynamic to 0, 3, 6 seems completely arbitrary and inconsistent. What's the reasoning for it?

Sorry, there is no specific explanation here for the meaning of (0, 3, 6). The 0 corresponds to second-level precision, 3 corresponds to millisecond precision, and 6 corresponds to nanosecond precision. These are currently the most commonly used precisions.

divanik replied:
> Making it dynamic to 0, 3, 6 seems completely arbitrary and inconsistent. What's the reasoning for it?

I don't fully agree that it's entirely arbitrary. It seems to relate to milliseconds and microseconds. From my understanding, we're trying to adjust this setting to align with Spark's writing format.

@alexey-milovidov, what are your thoughts? Do we need this feature, or would it add more complexity to the codebase than the potential benefits justify?

alexey-milovidov replied:
We need a different name for this setting.

`date_time_64_output_format_cut_trailing_zeros_align_to_groups_of_thousands`

lwz9103 commented:
I checked the failed tests, and all of them are related to test_storage_rabbitmq. The error message is `Error response from daemon: a prune operation is already running`. This seems to be unrelated to my PR. 
@divanik Could you help me take another look?

divanik commented:
> I checked the failed tests, and all of them are related to test_storage_rabbitmq. The error message is `Error response from daemon: a prune operation is already running`. This seems to be unrelated to my PR. @divanik Could you help me take another look?

Sometimes flaky tests just happen, it is possible to restart this tests. I will take a look on it when Alexey confirms that we need this change

alexey-milovidov commented on src/IO/WriteHelpers.h:
```diff
@@ -821,7 +821,23 @@ inline void writeDateTime64FractionalText(typename DecimalType::NativeType fract
         data[pos] += fractional % DateTime64(10);
 
-    writeString(&data[0], static_cast<size_t>(scale), buf);
+    if constexpr (trim_suffix_zeros)
```
The variable name should also be adjusted.

lwz9103 replied:
thanks, I will update later.

alexey-milovidov commented on src/IO/WriteHelpers.h:
```diff
@@ -934,7 +950,12 @@ inline void writeDateTimeText(time_t datetime, WriteBuffer & buf, const DateLUTI
+    char time_delimeter = ':',
+    char between_date_time_delimiter = ' ',
+    char fractional_time_delimiter = '.',
+    bool trim_suffix_zeros = false>
```
The variable name should also be adjusted.

lwz9103 replied:
thanks, I will update later.

alexey-milovidov commented on src/IO/tests/gtest_DateTimeToString.cpp:
```diff
@@ -212,3 +223,36 @@ INSTANTIATE_TEST_SUITE_P(DateTimeToString, DateTimeToStringParamTestDateTime64,
+             "1970-01-01 00:00:00.123456"
+         },
+    })
+ );
```
Add functional tests (.sql)

They should test for complex data types (such as Array of DateTime64) and various formats (such as JSON)

lwz9103 replied:
Thank you for the reminder. Your considerations are more comprehensive.

divanik reviewed (approved):
LGTM, I will deal with the flaky tests and merge

divanik commented:
Also I noticed that you are committing to the master of your fork. Please, next time create a new branch with command "git checkout -b <YOUR BRANCH NAME>" and commit to it. 
Also, don't forget to fix your master branch after merging this PR

lwz9103 commented on src/Functions/FunctionsConversion.cpp:
```diff
@@ -531,10 +532,13 @@ struct FormatImpl<DataTypeDateTime>
 struct FormatImpl<DataTypeDateTime64>
 {
-    template <typename ReturnType = void>
+    template <typename ReturnType = void, bool cut_trailing_zeros_align_to_groups_of_thousands = false>
```
Thank you for the reminder.

lwz9103 committed:
Improve compatibility of cast(timestamp as string) with spark

remove trailing zeros with the configuration date_time_64_output_format_cut_trailing_zeros_align_to_groups_of_thousands, which defaults to false to maintain consistency with previous behavior.
If config is true, output scale will be dynamic set in (0, 3, 6) according to tailing zeros. See changes in following examples.

- 2012-01-01 00:11:22.000000 -> 2012-01-01 00:11:22
- 2012-01-01 00:11:22.120000 -> 2012-01-01 00:11:22.120
- 2012-01-01 00:11:22.123400 -> 2012-01-01 00:11:22.123400

lwz9103 committed:
fix nullptr if query thread not initialized

lwz9103 commented:
> _This is an automated comment for commit [af941d0](https://github.com/ClickHouse/ClickHouse/commit/af941d00750b796dd136748683fddc7181de4f04) with description of existing statuses. It's updated for the latest CI running_
> 
> [❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69179/af941d00750b796dd136748683fddc7181de4f04/ci_running.html) to open a full report in a separate page
> 
> Check name	Description	Status
> AST fuzzer	Runs randomly generated queries to catch program errors. The build type is optionally given in parenthesis. If it fails, ask a maintainer for help	[❌ failure](https://s3.amazonaws.com/clickhouse-test-reports/69179/af941d00750b796dd136748683fddc7181de4f04/ast_fuzzer__debug_.html)
> Successful checks

Above error trace stack is as follows:

```
{0x18902074 <std::logic_error::logic_error(char const*)+52>,
0x70fa0c9 <std::out_of_range::out_of_range[abi:v15007](char const*)+9>, 
0x70fa07f <std::__1::__throw_out_of_range[abi:v15007](char const*)+31>, 
 
0x1105f64d, 0x119a58ad <DB::(anonymous namespace)::buildJoinClause(DB::ActionsDAG&, DB::ActionsDAG&, DB::ActionsDAG&, std::__1::shared_ptr<DB::PlannerContext> const&, std::__1::shared_ptr<DB::IQueryTreeNode> const&, std::__1::unordered_set<DB::IQueryTreeNode const*, std::__1::hash<DB::IQueryTreeNode const*>, std::__1::equal_to<DB::IQueryTreeNode const*>, std::__1::allocator<DB::IQueryTreeNode const*> > const&, std::__1::unordered_set<DB::IQueryTreeNode const*, std::__1::hash<DB::IQueryTreeNode const*>, std::__1::equal_to<DB::IQueryTreeNode const*>, std::__1::allocator<DB::IQueryTreeNode const*> > const&, DB::JoinNode const&, DB::JoinClause&)+3405>, 

0x1199fbba <DB::(anonymous namespace)::buildJoinClausesAndActions(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, DB::JoinNode const&, std::__1::shared_ptr<DB::PlannerContext> const&)+1914>, 

0x1199f312 <DB::buildJoinClausesAndActions(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IQueryTreeNode> const&, std::__1::shared_ptr<DB::PlannerContext> const&)+82>, 

0x1198b70f <DB::(anonymous namespace)::buildQueryPlanForJoinNode(std::__1::shared_ptr<DB::IQueryTreeNode> const&, DB::JoinTreeQueryPlan, DB::JoinTreeQueryPlan, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::PlannerContext>&)+655>, 

0x11981262 <DB::buildJoinTreeQueryPlan(std::__1::shared_ptr<DB::IQueryTreeNode> const&, DB::SelectQueryInfo const&, DB::SelectQueryOptions&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::PlannerContext>&)+9090>,

0x1196b82e <DB::Planner::buildPlanForQueryNode()+7534>, 
0x119685fe <DB::Planner::buildQueryPlanIfNeeded()+894>, 
0x119667ad <DB::InterpreterSelectQueryAnalyzer::getQueryPlan()+13>, 0x11c2ead6 <DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)+14166>, 0x11c2b05d <DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum)+189>, 
0x13021da9 <DB::TCPHandler::runImpl()+3945>, 0x13037599 <DB::TCPHandler::run()+25>, 0x1690c267 <Poco::Net::TCPServerConnection::start()+7>,
 0x1690c73e <Poco::Net::TCPServerDispatcher::run()+286>, 0x168b8cb2 <Poco::PooledThread::run()+178>, 0x168b69c3 <Poco::ThreadImpl::runnableEntry(void*)+131>, 0x7ff431cf8ac3, 0x7ff431d8a850, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}
```

I am not sure if this is related to my changes, so I need to run this test again.

baibaichen commented:
@Algunenano, would we merge it?

divanik committed:
Merge branch 'master' into master

divanik commented:
> > _This is an automated comment for commit [af941d0](https://github.com/ClickHouse/ClickHouse/commit/af941d00750b796dd136748683fddc7181de4f04) with description of existing statuses. It's updated for the latest CI running_
> > [❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69179/af941d00750b796dd136748683fddc7181de4f04/ci_running.html) to open a full report in a separate page
> > Check name	Description	Status
> > AST fuzzer	Runs randomly generated queries to catch program errors. The build type is optionally given in parenthesis. If it fails, ask a maintainer for help	[❌ failure](https://s3.amazonaws.com/clickhouse-test-reports/69179/af941d00750b796dd136748683fddc7181de4f04/ast_fuzzer__debug_.html)
> > Successful checks
> 
> Above error trace stack is as follows:
> 
> ```
> {0x18902074 <std::logic_error::logic_error(char const*)+52>,
> 0x70fa0c9 <std::out_of_range::out_of_range[abi:v15007](char const*)+9>, 
> 0x70fa07f <std::__1::__throw_out_of_range[abi:v15007](char const*)+31>, 
>  
> 0x1105f64d, 0x119a58ad <DB::(anonymous namespace)::buildJoinClause(DB::ActionsDAG&, DB::ActionsDAG&, DB::ActionsDAG&, std::__1::shared_ptr<DB::PlannerContext> const&, std::__1::shared_ptr<DB::IQueryTreeNode> const&, std::__1::unordered_set<DB::IQueryTreeNode const*, std::__1::hash<DB::IQueryTreeNode const*>, std::__1::equal_to<DB::IQueryTreeNode const*>, std::__1::allocator<DB::IQueryTreeNode const*> > const&, std::__1::unordered_set<DB::IQueryTreeNode const*, std::__1::hash<DB::IQueryTreeNode const*>, std::__1::equal_to<DB::IQueryTreeNode const*>, std::__1::allocator<DB::IQueryTreeNode const*> > const&, DB::JoinNode const&, DB::JoinClause&)+3405>, 
> 
> 0x1199fbba <DB::(anonymous namespace)::buildJoinClausesAndActions(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, DB::JoinNode const&, std::__1::shared_ptr<DB::PlannerContext> const&)+1914>, 
> 
> 0x1199f312 <DB::buildJoinClausesAndActions(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IQueryTreeNode> const&, std::__1::shared_ptr<DB::PlannerContext> const&)+82>, 
> 
> 0x1198b70f <DB::(anonymous namespace)::buildQueryPlanForJoinNode(std::__1::shared_ptr<DB::IQueryTreeNode> const&, DB::JoinTreeQueryPlan, DB::JoinTreeQueryPlan, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::PlannerContext>&)+655>, 
> 
> 0x11981262 <DB::buildJoinTreeQueryPlan(std::__1::shared_ptr<DB::IQueryTreeNode> const&, DB::SelectQueryInfo const&, DB::SelectQueryOptions&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::PlannerContext>&)+9090>,
> 
> 0x1196b82e <DB::Planner::buildPlanForQueryNode()+7534>, 
> 0x119685fe <DB::Planner::buildQueryPlanIfNeeded()+894>, 
> 0x119667ad <DB::InterpreterSelectQueryAnalyzer::getQueryPlan()+13>, 0x11c2ead6 <DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)+14166>, 0x11c2b05d <DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum)+189>, 
> 0x13021da9 <DB::TCPHandler::runImpl()+3945>, 0x13037599 <DB::TCPHandler::run()+25>, 0x1690c267 <Poco::Net::TCPServerConnection::start()+7>,
>  0x1690c73e <Poco::Net::TCPServerDispatcher::run()+286>, 0x168b8cb2 <Poco::PooledThread::run()+178>, 0x168b69c3 <Poco::ThreadImpl::runnableEntry(void*)+131>, 0x7ff431cf8ac3, 0x7ff431d8a850, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}
> ```
> 
> I am not sure if this is related to my changes. @divanik Can you help me confirm this issue, or tell me how to rerun this test individually?

The ideal policy should be the following:
we do not rerun failed tests

On practice:
we report all the failed tests to the repository issues
after that CH employer can rerun some of the failed jobs (a contributor can rerun only all the PR by pulling master or creating an empty commit, I do not like this fact but it is what it is). In your particular case it was a fuzzer issue, we do not rerun them at all (just try to make sure that it is not connected with your PR)

clickgapai commented:
Hi @lwz9103 @divanik — while reviewing this PR I found the following:

- **Bug (P1):** `writeDateTime64FractionalText` truncates significant digits for scale > 6 — https://github.com/ClickHouse/ClickHouse/issues/104997

Happy to discuss — close anything that's wrong or already addressed.
